File: bdp-devops/ambari-versions/ambari-3.0.0.0-SNAPSHOT/ambari-server/src/main/java/org/apache/ambari/server/topology/addservice/Host.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ambari.server.topology.addservice;

import java.util.Objects;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import io.swagger.annotations.ApiModelProperty;

public final class Host {

  private static final String FQDN = "fqdn";

  private final String fqdn;

  @JsonCreator
  public Host(@JsonProperty(FQDN) String fqdn) {
    this.fqdn = fqdn;
  }

  @JsonProperty(FQDN)
  @ApiModelProperty(name = FQDN)
  public String getFqdn() {
    return fqdn;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    Host other = (Host) o;

    return Objects.equals(fqdn, other.fqdn);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(fqdn);
  }

  @Override
  public String toString() {
    return "host: " + fqdn;
  }

}

File: olap/ByConity/src/Functions/host.cpp
/*
 * Copyright (2022) Bytedance Ltd. and/or its affiliates
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <Functions/IFunction.h>
#include <Functions/FunctionFactory.h>
#include <DataTypes/DataTypeString.h>
#include <Common/HostWithPorts.h>
#include <common/getFQDNOrHostName.h>
#include <Core/Field.h>


namespace DB
{

/// Get the host. It is constant on single server, but is not constant in distributed queries.
class FunctionHost : public IFunction
{
public:
    static constexpr auto name = "host";
    static FunctionPtr create(ContextPtr)
    {
        return std::make_shared<FunctionHost>();
    }

    String getName() const override
    {
        return name;
    }

    bool isDeterministic() const override { return false; }

    bool isDeterministicInScopeOfQuery() const override
    {
        return false;
    }

    size_t getNumberOfArguments() const override
    {
        return 0;
    }

    DataTypePtr getReturnTypeImpl(const DataTypes & /*arguments*/) const override
    {
        return std::make_shared<DataTypeString>();
    }

    /** convertToFullColumn needed because in distributed query processing,
      *    each server returns its own value.
      */
    ColumnPtr executeImpl(const ColumnsWithTypeAndName &, const DataTypePtr & result_type, size_t input_rows_count) const override
    {
        return result_type->createColumnConst(input_rows_count, getHostIPFromEnv())->convertToFullColumnIfConst();
    }
};


void registerFunctionHost(FunctionFactory & factory)
{
    factory.registerFunction<FunctionHost>();
}

}
